packages/core/src/stateUtils.ts

```diff
@@ -1805,143 +1605,185 @@ export function resolveActionsAndContext<
-        machine,
-        eventlessTransitions,
-        maybeNextState,
-        predictableExec,
-        maybeNextState._event
+    }
+
+    if (enabledTransitions.length) {
+      const currentActions = nextState.actions;
+      nextState = microstep(
+        enabledTransitions,
+        nextState,
+        actorCtx,
+        nextState._event
       );
-      _internalQueue.push(...maybeNextState._internalQueue);
-      maybeNextState.actions.unshift(...currentActions);
+      nextState.actions.unshift(...currentActions);
+
+      states.push(nextState);
     }
   }
 
-  // Add tags
-  maybeNextState.tags = new Set(
-    flatten(maybeNextState.configuration.map((sn) => sn.tags))
+  if (nextState.done) {
+    // Perform the stop step to ensure that child actors are stopped
+    _stopStep(nextState._event, nextState, actorCtx);
+  }
+
+  return {
+    state: nextState,
+    microstates: states
+  };
+}
+
+function _stopStep(
+  scxmlEvent: SCXML.Event<any>,
+  nextState: AnyState,
+  actorCtx: ActorContext<any, any> | undefined
+): AnyState {
+  const stoppedState = cloneState(nextState, {
+    _event: scxmlEvent,
+    actions: []
+  });
+
+  stoppedState.actions.length = 0;
+
+  for (const stateNode of nextState.configuration.sort(
+    (a, b) => b.order - a.order
+  )) {
+    stoppedState.actions.push(...stateNode.definition.exit);
+  }
+
+  for (const child of Object.values(nextState.children)) {
+    stoppedState.actions.push(stop(child));
+  }
+
+  const { actions, context } = resolveActionsAndContext(
+    stoppedState.actions,
+    stoppedState._event,
+    stoppedState,
+    // stoppedState.context,
+    actorCtx
   );
 
-  maybeNextState.machine = machine;
+  stoppedState.actions = actions;
+  stoppedState.context = context;
 
-  return maybeNextState;
+  return stoppedState;
 }
 
-function resolveHistoryValue<
-  TContext extends MachineContext,
-  TEvent extends EventObject
->(
-  currentState: State<TContext, TEvent> | undefined,
-  exitSet: Array<StateNode<TContext, TEvent>>
-): HistoryValue<TContext, TEvent> {
-  const historyValue: Record<
-    string,
-    Array<StateNode<TContext, TEvent>>
-  > = currentState ? currentState.historyValue : {};
-  if (currentState && currentState.configuration) {
-    // From SCXML algorithm: https://www.w3.org/TR/scxml/#exitStates
-    for (const exitStateNode of exitSet) {
-      for (const historyNode of getHistoryNodes(exitStateNode)) {
-        let predicate: (sn: StateNode<TContext, TEvent>) => boolean;
-        if (historyNode.history === 'deep') {
-          predicate = (sn) =>
-            isAtomicStateNode(sn) && isDescendant(sn, exitStateNode);
-        } else {
-          predicate = (sn) => {
-            return sn.parent === exitStateNode;
-          };
+function _selectTransitions(
+  scxmlEvent: SCXML.Event<any>,
+  nextState: AnyState
+): AnyTransitionDefinition[] {
+  return nextState.machine.getTransitionData(nextState, scxmlEvent);
+}
+
+function _selectEventlessTransitions(
```

Comment:
this is a weird convention (`_`) for module-level function - if we want to have it "private" we just have to keep it local, non-exported and that's it

Reply:
Makes sense